activerecord/lib/active_record/query_logs.rb

```diff
@@ -0,0 +1,187 @@
+    thread_mattr_accessor :cached_comment, instance_accessor: false
+
+    class << self
+      # Updates the context used to construct tags in the SQL comment.
+      # Resets the cached comment if <tt>cache_query_log_tags</tt> is +true+.
+      def update_context(**options)
+        context.merge!(**options.symbolize_keys)
+        self.cached_comment = nil
+      end
+
+      # Updates the context used to construct tags in the SQL comment during
+      # execution of the provided block. Resets provided values to nil after
+      # the block is executed.
+      def set_context(**options)
+        update_context(**options)
+        yield if block_given?
+      ensure
+        update_context(**options.transform_values! { nil })
+      end
+
+      # Temporarily tag any query executed within `&block`. Can be nested.
+      def with_tag(tag, &block)
+        inline_tags.push(tag)
+        yield if block_given?
+      ensure
+        inline_tags.pop
+      end
+
+      def add_query_log_tags_to_sql(sql) # :nodoc:
+        comments.each do |comment|
+          unless sql.include?(comment)
+            sql = prepend_comment ? "#{comment} #{sql}" : "#{sql} #{comment}"
+          end
+        end
+        sql
+      end
+
+      private
+        # Returns an array of comments which need to be added to the query, comprised
+        # of configured and inline tags.
+        def comments
+          [ comment, inline_comment ].compact
+        end
+
+        # Returns an SQL comment +String+ containing the query log tags.
+        # Sets and returns a cached comment if <tt>cache_query_log_tags</tt> is +true+.
+        def comment
+          if cache_query_log_tags
+            self.cached_comment ||= uncached_comment
+          else
+            uncached_comment
+          end
+        end
+
+        def uncached_comment
+          content = tag_content
+          if content.present?
+            "/*#{escape_sql_comment(content)}*/"
+          end
+        end
+
+        # Returns a +String+ containing any inline comments from +with_tag+.
+        def inline_comment
+          return nil unless inline_tags.present?
+          "/*#{escape_sql_comment(inline_tag_content)}*/"
+        end
+
+        # Return the set of active inline tags from +with_tag+.
+        def inline_tags
+          context[:inline_tags] ||= []
+        end
+
+        def context
+          Thread.current[:active_record_query_log_tags_context] ||= {}
+        end
+
+        def escape_sql_comment(content)
+          content.to_s.gsub(%r{ (/ (?: | \g<1>) \*) \+? \s* | \s* (\* (?: | \g<2>) /) }x, "")
+        end
+
+        def tag_content
+          tags.flat_map { |i| [*i] }.filter_map do |tag|
+            key, value_input = tag
+            val = case value_input
+                  when nil then instance_exec(&taggings[key]) if taggings.has_key? key
+                  when Proc then instance_exec(&value_input)
```

Comment:
So I'm trying to update our app to replace Marginalia, and I'm having a bit of trouble with this `instance_exec`. My main issue it that sorbet really doesn't like it, and that's on us to fix, but ultimately I wonder if `value_input.call(context)` wouldn't be better.

What do you think?
First it would avoid exposing all private methods as an API, and also it would allow passing anything that responds to `call`, not just a proc.

Reply:
Seems fine to me 👍

Reply:
I'll PR it.